### ISSUE TITLE
Hallucianating Questionaire

### ISSUE THREAD

**Issue by rbrox** [komalioruganti/HydroMetrics#1]
## The observed problems are:
1. On the question `Is your household using any water purification methods, such as filters or purifiers`, The tf values seem to be spiking upon repetative clicking of the yes button

2. Going back to older questions in the form seems to be causing the form to hallucinate and show wrong values for the tf.

<img width="1483" alt="Screenshot 2024-03-16 at 12 52 40 PM" src="https://github.com/komalioruganti/HydroMetrics/assets/111554368/15a6c890-ffcb-4258-b8d0-3f03f97a607b">

<img width="1139" alt="Screenshot 2024-03-16 at 12 56 13 PM" src="https://github.com/komalioruganti/HydroMetrics/assets/111554368/9af5c67a-5f1e-4365-9d70-3769d4065b8b">

## My understanding of the current code

Loc: `HydroMetrics/client/src/Pages/q.js`

### States used:

```javascript
/**
 * @param {list} answers - Selected answer object
 * @param {number} members - Count of members in family (numeric)
 * @param {number} tf - Total footprint (water footprint) (numeric)
 * @param {*} prev - Previous value (unknown type)
 */
```

### Current code Deign/Setup

```javascript
// Imports

// State Declarations
// You may place your state declarations here

/**
 * @description Handles form submission.
 * Most probably hitting the backend with the form data.
 */
function handleSubmit() {
  // Your implementation for handling form submission goes here
}

/**
 * @description Handles input change.
 * This function is called after every question.
 *
 * @param {object} value - The selected/answer
 * @param {string} q - The question number
 * @returns {number} - Returns 500, doesn't return much, the objective is to change the state variable
 */
function handleInputChange(value, q) {
  // Your implementation for handling input change goes here
}

return <>All questionaire elements and the visual effects</>;
```

---

# Proposed Solution

1. The answers list can be converted to a dictionary with the question number as the key and the answer as the value. This will make it easier to handle the state changes and also make the code more readable.

```javascript
const [answers, setAnswers] = useState({
  members: 0,
  question0: false,
  question1: false,
  question2: false,
  question3: 0,
  question4: 0,
  question5: 0,
  question6: 0,
  question7: 0,
  question8: 0,
});
```

2. A computeTf function can be used to calculate the tf value based on the answers dictionary. This function can be called after every state change to update the tf value.

```javascript
/**
 * @description - This function is used to calculate the tf value based on the answers dictionary
 * @returns {number} - The tf value
 * @param {object} answers - The answers dictionary
 *
*/
  const computeTf = () => {
	let tf = 0;
	if (answers.question0) {
	 ...
	}
	if (answers.question1) {
	 ...
	}
	if (answers.question2) {
	 ...
	}
	...
	setTf(tf);
  };
```

If you find this makes sense ill go ahead and implement the changes. If I seem to be missing something, please let me know.

**Comment by komalioruganti** [comment]
Yes go ahead but make sure the water footprint value is calculated based on the formula properly. If the formula goes wrong the values will start going haywire. The total water footprint value should also be displayed correctly. So please keep that in mind but yes you can go ahead.

**Comment by rbrox** [comment]
Cool then

**Comment by rbrox** [comment]
Progress until now the questionaire is long

https://github.com/komalioruganti/HydroMetrics/assets/111554368/07b4b2a9-057b-4f8a-9987-d0713b8dc25d

**Comment by komalioruganti** [comment]
Everything works fine but looks like an error is popping up at the end.

**Comment by rbrox** [comment]
Yup the last part wasnt fixed
 fixed till the vehicles question, (will get it done by tmmrw)


https://github.com/komalioruganti/HydroMetrics/assets/111554368/3fd14454-f4ea-4587-aa45-48028c80deab
